PLT-7591: Improved Auditor Report Upload

EehMauro opened the pull request:
@coderabbitai: ignore

amnambiar commented:
Note: we need to integrate changes from PR #8 here once that is approved/merged

RSoulatIOHK commented:
Love the new design.

### Error messages
Ok, I don't know why, but at first, the checks for the values were not happening until I clicked on Send report. And after that it would only show me one field that was wrong. I fixed them in the order requested:

- Auditor Information
    - Name 
    - Website
    - Logo

And then, it showed all the other field with wrong values, and will display live the error messages of everything I filled wrong.

### Wrong Regex
I don't get from the code why this Website value (example.com or www.example.com) is not ok for the regex:
![image](https://github.com/input-output-hk/dapps-certification-web/assets/117812549/a9614432-1eb7-414c-b158-e95462fa982b)

EehMauro commented:
> Love the new design.
> 
> ### Error messages
> Ok, I don't know why, but at first, the checks for the values were not happening until I clicked on Send report. And after that it would only show me one field that was wrong. I fixed them in the order requested:
> 
> * Auditor Information
>   
>   * Name
>   * Website
>   * Logo
> 
> And then, it showed all the other field with wrong values, and will display live the error messages of everything I filled wrong.
> 
> ### Wrong Regex
> I don't get from the code why this Website value (example.com or [www.example.com](http://www.example.com)) is not ok for the regex: ![image](https://user-images.githubusercontent.com/117812549/269223878-a9614432-1eb7-414c-b158-e95462fa982b.png)

Fixed!

EehMauro commented:
Added Auditor Report Upload fields prefill with user profile

amnambiar commented:
@EehMauro , Here are my thoughts -
'Auditor Report Upload' and ['Review Certification Metadata'](PR #8) are same just with Subject and Report URLs missing in the later. Rest all schema, functionalities are the same. Hence I created a common MetadataForm component to share between the two; by supplying separate configurations to build the form, and schemas (see PR #20 for latest).

Have you kept this in mind in this PR, otherwise what is your suggestion. We need to discuss over this, to reduce redundancy and make this simpler.

amnambiar commented:
@EehMauro , Also I see you are not making use of schema validators like yup. Is there a particular reason why? Also, do you advocate in removing them throughout the app?

amnambiar commented:
> @EehMauro , Here are my thoughts - 'Auditor Report Upload' and ['Review Certification Metadata'](PR #8) are same just with Subject and Report URLs missing in the later. Rest all schema, functionalities are the same. Hence I created a common MetadataForm component to share between the two; by supplying separate configurations to build the form, and schemas (see PR #20 for latest).
> 
> Have you kept this in mind in this PR, otherwise what is your suggestion. We need to discuss over this, to reduce redundancy and make this simpler.

PR #8 was merged after a lot of integration testing / bug fixes / schema corrections :laughing: 
If we redo the component entirely in this PR , wouldn't we have to apply the same effort here as well? I'm a bit sceptical with this :upside_down_face:

EehMauro commented:
Hi @amnambiar, I can merge #20 here so I can re-use the form you made and the yup validations

Commit:
Bulk commit

Commit:
Fix styles

RSoulatIOHK commented on src/compositions/InputGroup/utils/schemas.ts:
```diff
@@ -0,0 +1,67 @@
+
+export const WalletAddress = yup.string()
+  .matches(/^(addr_test1|addr1)[a-zA-Z0-9]{53,}$/, {
+    message: 'The field value must be a valid wallet address',
```
```suggestion
    message: 'The field value must be a valid contract address',
```

RSoulatIOHK commented on src/utils/utils.ts:
```diff
@@ -86,6 +86,11 @@ export const transformEmptyStringToNullInObj = (obj: any) => {
 }
 
+export const removeEmptyStringsDeep = (obj: any) => {
+  Object.keys(obj).forEach(key => (obj[key] && typeof obj[key] === 'object') && removeEmptyStringsDeep(obj[key]) || (obj[key] === '') && delete obj[key]);
```
The logic is probably correct but it's kind of hard to follow, could you add some parantheses to better separate the `and`s and `or`s

RSoulatIOHK commented on src/pages/session/index.tsx:
```diff
@@ -27,7 +31,7 @@ export default () => {
   }, [hasAnActiveSubscription, isSessionFetched, location.pathname]);
 
-  if (!isSessionFetched) {
+  if (!isSessionFetched || hasAnActiveSubscription && profile === null) {
```
Could you add some parantheses to make it easier to follow the logic?

RSoulatIOHK commented on src/compositions/InputGroup/utils/index.ts:
```diff
@@ -0,0 +1,46 @@
+  for (const field of fields) {
+    if (field.type == FieldType.Object) {
+      schema[field.name] = buildSchema(field.fields || []);
+    } else if (field.type == FieldType.Array) {
```
I don't really know TS so sorry if this is a dumb question, but wouldn't it be better to use "==="?

RSoulatIOHK commented on src/compositions/InputGroup/utils/index.ts:
```diff
@@ -0,0 +1,46 @@
+const buildSchema = (fields: Field[]): AnyObjectSchema => {
+  const schema = {} as {[key: string]: AnySchema};
+  for (const field of fields) {
+    if (field.type == FieldType.Object) {
```
I don't really know TS so sorry if this is a dumb question, but wouldn't it be better to use "==="?

RSoulatIOHK reviewed (changes requested):
Two more things but I don't know where to point it into the code but:

- The checks on the values don't start until I click a first time on "Send report"
- If I put a too long "Script Hash" like : `01234567890123456789012345678901234567890123456789012345678901234567890012345678901234567890123456789012345678901234567890123456789012345678900123456789012345678901234567890123456789012345678901234567890123456789001234567890123456789012345678901234567890123456789012345678901234567890` it never shows on the form that it is wrong but I get an error when clicking on "Send report"

RSoulatIOHK commented on src/compositions/InputGroup/utils/schemas.ts:
```diff
@@ -0,0 +1,67 @@
+export const Email = yup.string()
+  .email('The field value must be a valid email address');
+
+export const Subject = yup.string()
```
Subject is more complex than that:
"Subject: string
pattern: ^[A-Za-z0-9_]{1,64}$"

RSoulatIOHK reviewed (changes requested):
Sorry, Missed this during the first review

Commit:
Resolved comments

amnambiar commented:
1. reportURLs are to be sent as array of strings in report.reportURLs in the payload
2. Move Add Script to after all the the DApp Scripts (like Romain mentioned)
3. Error in form submit is retained while navigating away and coming back, even though the form is cleared. Clear errors in the form, while moving away from the page.

@EehMauro bugs/changes to be addressed here

Commit:
Resolved comments

amnambiar commented:
@EehMauro , Pardon my bad. Can you please fix (1)

POST /auditor/reports expects . Can you fix that. And, the error `400 user error (Report URLs do not match)` when you give urls to different PDFs are because they are basically NOT different urls to the same report file(expected). 
```
"report": [
    "string"
  ],
```

Commit:
Fixed report upload schema

amnambiar commented:
@EehMauro Bugs - In Review Certification Metadata -- 
1. remove fields Certification Level, Subject, Report URLs
2. Add a cancel or close button to cancel the modal if he chooses to, as this is not mandatory

amnambiar commented on src/pages/certification/Certification.tsx:
```diff
@@ -94,13 +94,11 @@ const Certification = () => {
-                        ? "Run: " + ellipsizeString(runId)
-                        : "Testing Tool"}
-                </h2>
+        <Container className="py-4" maxWidth="xl">
```
@EehMauro ,

this causes a max-width to content-section which doesn't actually expand the content to fit the screen when >4k. We don't want <Container> elements for that reason. Atleast in Testing and History views. 

I am reverting this change here, as it also breaks the button styles (Test again, New test)

Otherwise make sure to use  maxWidth={false} like in FullReportTable.tsx

amnambiar committed:
PLT-7591 Fixed Container and maxwidth in Testing and History

Commit:
Merge branch 'master' into PLT-7591

Commit:
Resolved comments

amnambiar committed:
Merge branch 'master' of https://github.com/input-output-hk/dapps-certification-web into PLT-7591

amnambiar committed:
PLT-7047  Fixed githubToken:null if empty in /profile/current api

amnambiar commented:
>  In Review Certification Metadata --
> 
> remove fields Certification Level, Subject, Report URLs
> Add a cancel or close button to cancel the modal if he chooses to, as this is not mandatory

This is done.

amnambiar reviewed (approved):
LGTM now
